### PR TITLE
WIP - Liq 2.0 Upgrade

### PR DESCRIPTION
**Pull request by brianmcmichael**
WIP

Needs more tests, data optimizations, and some team discussion about design. Not sure if I love the design, but should give us a little more flexibility with new asset types.

### CODE REVIEW

File: src/IlkRegistry.sol [hunk 1]
```diff
@@ -85,29 +96,33 @@ contract IlkRegistry {
...
     struct Ilk {
-        uint256 pos;   // Index in ilks array
-        address gem;   // The token contract
-        address pip;   // Token price
-        address join;  // DSS GemJoin adapter
-        address flip;  // Auction contract
-        uint256 dec;   // Token decimals
-        string name;   // Token name
-        string symbol; // Token symbol
+        uint128 pos;    // Index in ilks array
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:103]: Might be worth thinking about how to better pack this struct.

File: src/IlkRegistry.sol [hunk 2]
```diff
@@ -137,10 +152,23 @@ contract IlkRegistry {
         (address _pip,) = spot.ilks(_ilk);
         require(_pip != address(0), "IlkRegistry/pip-invalid");
 
-        (address _flip,,) = cat.ilks(_ilk);
-        require(_flip != address(0), "IlkRegistry/flip-invalid");
-        require(FlipLike(_flip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
-        require(FlipLike(_flip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
+        (address _clip,,,) = dog.ilks(_ilk);
+        (address _flip,,)  = cat.ilks(_ilk);
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:156]: Maybe do this in the else block.

File: src/IlkRegistry.sol [hunk 3]
```diff
@@ -137,10 +152,23 @@ contract IlkRegistry {
...
+            _xlip = _clip;
+            _code = 1;
+        } else if (_flip != address(0)) {
+            require(FlipLike(_flip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
+            require(FlipLike(_flip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
+            _xlip = _flip;
+            _code = 2;
+        } else {
+            revert("IlkRegistry/invalid-auction-contract");
+        }
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:172]: Maybe check that xlip is authed on the vat?

File: src/IlkRegistry.sol [hunk 4]
```diff
@@ -332,16 +370,68 @@ contract IlkRegistry {
...
+            uint128 code,
+            address gem,
+            address pip,
+            address join,
+            address xlip,
+            uint256 dec,
+            string calldata name,
+            string calldata symbol)
+        external auth {
+            require(code != 0, "IlkRegistry/invalid-type");
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:410]: Also ensure `code < uint128(-1)`

File: src/IlkRegistry.sol [hunk 5]
```diff
@@ -85,29 +96,33 @@ contract IlkRegistry {
...
-        uint256 pos;   // Index in ilks array
-        address gem;   // The token contract
-        address pip;   // Token price
-        address join;  // DSS GemJoin adapter
-        address flip;  // Auction contract
-        uint256 dec;   // Token decimals
-        string name;   // Token name
-        string symbol; // Token symbol
+        uint128 pos;    // Index in ilks array
+        uint128 code;   // Class code (1 - clip, 2 - flip, 3+ - other)
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:104]: maybe call this `class`

File: src/IlkRegistry.sol [hunk 6]
```diff
@@ -137,10 +152,23 @@ contract IlkRegistry {
...
 
-        (address _flip,,) = cat.ilks(_ilk);
-        require(_flip != address(0), "IlkRegistry/flip-invalid");
-        require(FlipLike(_flip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
-        require(FlipLike(_flip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
+        (address _clip,,,) = dog.ilks(_ilk);
+        (address _flip,,)  = cat.ilks(_ilk);
+        address  _xlip;
+        uint128  _code;
+        if (_clip != address(0)) {
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:159]: Maybe consider a path where we want to revert from a clip back to a flip.

File: src/IlkRegistry.sol [hunk 7]
```diff
@@ -332,16 +370,68 @@ contract IlkRegistry {
         (address _pip,) = spot.ilks(ilk);
         require(_pip != address(0), "IlkRegistry/pip-invalid");
 
-        (address _flip,,) = cat.ilks(ilk);
-        require(_flip != address(0), "IlkRegistry/flip-invalid");
-        require(FlipLike(_flip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
-        require(FlipLike(_flip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
+        (address _clip,,,) = dog.ilks(ilk);
```
Comment (brianmcmichael) [inline src/IlkRegistry.sol:373]: Maybe the public update function should just be removed entirely, since it can override governance-induced updates...

Comment (brianmcmichael) [reply src/IlkRegistry.sol:373]: Or we keep ability to update pip and not [f/c]lip

File: src/IlkRegistry.sol (hunk 3 above)
Comment (gbalabasquer) [reply src/IlkRegistry.sol:172]: I guess we can take care of that in the spell.

**Comment by gbalabasquer** [comment]
Maybe we want to mark as `indexed` the values of the events.

File: src/IlkRegistry.sol [hunk 8]
```diff
@@ -85,29 +96,33 @@ contract IlkRegistry {
...
+        string  name;   // Token name
+        string  symbol; // Token symbol
     }
 
     mapping (bytes32 => Ilk) public ilkData;
     bytes32[] ilks;
 
     // Initialize the registry
-    constructor(address vat_, address cat_, address spot_) public {
+    constructor(address vat_, address dog_, address cat_, address spot_) public {
```
Comment (gbalabasquer) [inline src/IlkRegistry.sol:118]: Super super nit: we might want to keep the `cat` - `dog` order everywhere.

File: src/IlkRegistry.sol [hunk 9]
```diff
@@ -137,10 +152,23 @@ contract IlkRegistry {
...
+            _xlip = _clip;
+            _class = 1;
+        } else {
+            (address _flip,,)  = cat.ilks(_ilk);
+            require(_flip != address(0), "IlkRegistry/invalid-auction-contract");
+            require(FlipLike(_flip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
+            require(FlipLike(_flip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
+            _xlip = _flip;
+            _class = 2;
+        }
```
Comment (gbalabasquer) [inline src/IlkRegistry.sol:171]:
```suggestion
        (address _xlip,,,) = dog.ilks(_ilk);

        uint96  _class = 1;
        if (_clip != address(0)) {
            require(ClipLike(_clip).dog() == address(dog), "IlkRegistry/clip-wrong-dog");
            require(ClipLike(_clip).vat() == address(vat), "IlkRegistry/clip-wrong-vat");
        } else {
            (address _xlip,,)  = cat.ilks(_ilk);
            require(_xlip != address(0), "IlkRegistry/invalid-auction-contract");
            require(FlipLike(_xlip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
            require(FlipLike(_xlip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
            _class = 2;
        }
```

Comment (gbalabasquer) [reply src/IlkRegistry.sol:171]: A bit shorter version, however yours is a bit more explicit. So feel free to take it or not.

File: src/IlkRegistry.sol [hunk 10]
```diff
@@ -332,16 +374,50 @@ contract IlkRegistry {
...
         require(_pip != address(0), "IlkRegistry/pip-invalid");
 
-        (address _flip,,) = cat.ilks(ilk);
-        require(_flip != address(0), "IlkRegistry/flip-invalid");
-        require(FlipLike(_flip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
-        require(FlipLike(_flip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
-
-        ilkData[ilk].pip   = _pip;
-        ilkData[ilk].flip  = _flip;
+        ilkData[ilk].pip    = _pip;
```
Comment (gbalabasquer) [inline src/IlkRegistry.sol:377]: Don't we want to repeat the code here as well to update the `xlip`?

**Comment by gbalabasquer** [comment]
Some comments and suggestions added, but so far the new structure looks good to me.

File: src/IlkRegistry.sol (hunk 3 above)
Comment (brianmcmichael) [reply src/IlkRegistry.sol:172]: The `add` function is still unauthed here, so we need the sanity checks to ensure an invalid ilk isn't getting in.

File: src/IlkRegistry.sol (hunk 8 above)
Comment (brianmcmichael) [reply src/IlkRegistry.sol:118]: Yeah, some things got moved around to optimize for clipper collaterals. I'll make the update.

File: src/IlkRegistry.sol [hunk 11]
```diff
@@ -328,25 +371,62 @@ contract IlkRegistry {
     function update(bytes32 ilk) external {
         require(JoinLike(ilkData[ilk].join).vat() == address(vat), "IlkRegistry/invalid-ilk");
         require(JoinLike(ilkData[ilk].join).live() == 1, "IlkRegistry/ilk-not-live-use-remove-instead");
+        uint96 _class = ilkData[ilk].class;
+        require(_class == 1 || _class == 2, "IlkRegistry/invalid-class");
```
Comment (kmbarry1) [inline src/IlkRegistry.sol:375]: So, even though we know RWA is coming soon and we plan to use 3+ for such ilks, we'll just plan to redeloy the registry? The alternative would be some sort of "addClass" privileged functionality that only the pause proxy can call.

Comment (kmbarry1) [reply src/IlkRegistry.sol:375]: Ah nvm, I see the function below provides an equivalent ability.

**Comment by kmbarry1** [comment]
I'm not as familiar with this as with other things, but no major objections from me on the new structure.

File: Makefile [hunk 12]
```diff
@@ -1,5 +1,5 @@
-all    :; SOLC_FLAGS="--optimize --optimize-runs=1000000" dapp --use solc:0.6.7 build
+all    :; DAPP_BUILD_OPTIMIZE=1 DAPP_BUILD_OPTIMIZE_RUNS=1000000 dapp --use solc:0.6.12 build
 clean  :; dapp clean
-test   :; SOLC_FLAGS="--optimize --optimize-runs=1000000" dapp --use solc:0.6.7 test -v
-deploy-mainnet :; SOLC_FLAGS="--optimize --optimize-runs=1000000" dapp --use solc:0.6.7 build && dapp create IlkRegistry 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B 0xa5679C04fc3d9d8b0AaB1F0ab83555b301cA70Ea 0x65C79fcB50Ca1594B025960e539eD7A9a6D434A3
-deploy-kovan :; SOLC_FLAGS="--optimize --optimize-runs=1000000" dapp --use solc:0.6.7 build && dapp create IlkRegistry 0xbA987bDB501d131f766fEe8180Da5d81b34b69d9 0xdDb5F7A3A5558b9a6a1f3382BD75E2268d1c6958 0x3a042de6413eDB15F2784f2f97cC68C7E9750b2D
+test   :; DAPP_BUILD_OPTIMIZE=1 DAPP_BUILD_OPTIMIZE_RUNS=1000000 dapp --use solc:0.6.12 test -v
+deploy-mainnet :; make && dapp create IlkRegistry 0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B 0xa5679C04fc3d9d8b0AaB1F0ab83555b301cA70Ea 0x65C79fcB50Ca1594B025960e539eD7A9a6D434A3
```
Comment (brianmcmichael) [inline Makefile:4]: Needs update once we have a dog address.

File: src/IlkRegistry.sol [hunk 13]
```diff
@@ -123,27 +137,36 @@ contract IlkRegistry {
...
+        if (_xlip != address(0)) {
+            require(ClipLike(_xlip).dog() == address(dog), "IlkRegistry/clip-wrong-dog");
+            require(ClipLike(_xlip).vat() == address(vat), "IlkRegistry/clip-wrong-vat");
+        } else {
+            (_xlip,,)  = cat.ilks(_ilk);
+            require(_xlip != address(0), "IlkRegistry/invalid-auction-contract");
+            require(FlipLike(_xlip).cat() == address(cat), "IlkRegistry/flip-wrong-cat");
+            require(FlipLike(_xlip).vat() == address(vat), "IlkRegistry/flip-wrong-vat");
+            _class = 2;
+        }
```
Comment (gbalabasquer) [inline src/IlkRegistry.sol:166]:
```suggestion
        if (_xlip == address(0)) {
            (_xlip,,)  = cat.ilks(_ilk);
            require(_xlip != address(0), "IlkRegistry/invalid-auction-contract");
            _class = 2;
        }
```

**Comment by godsflaw** [comment]
@ yaronvel ping.  There are a few changes to the interface on the registry.  Look at `info()` return values and the removal of `flip()` for examples of how integrations might break with this change.  Let us know if there's anything less disruptive you would suggest.

**Comment by yaronvel** [comment]
> @yaronvel ping. There are a few changes to the interface on the registry. Look at `info()` return values and the removal of `flip()` for examples of how integrations might break with this change. Let us know if there's anything less disruptive you would suggest.

Thanks for the heads up.
In B.Protocol we only use the chainlog contract. We don't use the ilk registry.

It seems like you are preparing to a situation where the dog and cat will co-exist for a while. 
If this is indeed the case, it would be nice to have a function at the ilk registry that returns chop value per ilk.

> @yaronvel ping. There are a few changes to the interface on the registry. Look at `info()` return values and the removal of `flip()` for examples of how integrations might break with this change. Let us know if there's anything less disruptive you would suggest.

> @yaronvel ping. There are a few changes to the interface on the registry. Look at `info()` return values and the removal of `flip()` for examples of how integrations might break with this change. Let us know if there's anything less disruptive you would suggest.

Thanks for the heads up.
In B.Protocol we only use the chainlog contract. We don't use the ilk registry.

It seems like you are preparing to a situation where the dog and cat will co-exist for a while. 
If this is indeed the case, it would be nice to have a function at the ilk registry that returns chop value per ilk.

**Comment by godsflaw** [comment]
> > @yaronvel ping. There are a few changes to the interface on the registry. Look at `info()` return values and the removal of `flip()` for examples of how integrations might break with this change. Let us know if there's anything less disruptive you would suggest.
> 
> Thanks for the heads up.
> In B.Protocol we only use the chainlog contract. We don't use the ilk registry.
> 
> It seems like you are preparing to a situation where the dog and cat will co-exist for a while.
> If this is indeed the case, it would be nice to have a function at the ilk registry that returns chop value per ilk.

You will have the ability to check `class` on the ilk-registry to check if it's LIQ-1.2 vs. LIQ-2.x.  From there, you can grab the `cat` or `dog` and get at the `chop`.  Will this work for you?

**Review by godsflaw — APPROVED** [review]
LGTM

**Review by e18r — APPROVED** [review]
LGTM

**Pull request merged**